File: testdata/fixtures/kotlin/CWE-78b.kt
// CWE-78: OS command injection via Runtime.exec with user input
package fixture
class Test78b {
    fun vuln(host: String) {
        // ruleid: kotlin-os-command
        Runtime.getRuntime().exec(host)
        // ruleid: kotlin-os-command
        ProcessBuilder(host, "arg").start()
    }
    fun vuln2(cmd: String) {
        // ruleid: kotlin-os-command
        Runtime.getRuntime().exec(cmd)
        // ruleid: kotlin-os-command
        ProcessBuilder(cmd).start()
    }
    fun safe() {
        // ok: kotlin-os-command
        Runtime.getRuntime().exec("ls")
        // ok: kotlin-os-command
        ProcessBuilder("echo", "hi").start()
    }
    fun safe2() {
        // ok: kotlin-os-command
        Runtime.getRuntime().exec("echo fixed")
    }
}

File: testdata/fixtures/csharp/CWE-78b.cs
// CWE-78: OS command injection — untrusted input reaches process sink.
// Request.Query["cmd"] flows to Process.Start(cmd) without validation.
// The rule flags those flows; fix by mapping to allow-listed commands
// and never passing raw input.
using System.Diagnostics;
using Microsoft.AspNetCore.Mvc;

namespace Fixture {
    public class OsCommandController : Controller {
        public IActionResult FromQuery() {
            string cmd = Request.Query["cmd"];
            // ruleid: csharp-os-command-injection
            Process.Start(cmd);
            return Ok();
        }

        public IActionResult FromForm(string tool) {
            // ruleid: csharp-os-command-injection
            Process.Start(tool);
            return Ok();
        }

        public IActionResult WithArgs() {
            string arg = Request.Form["arg"];
            // ruleid: csharp-os-command-injection
            Process.Start("ping", arg);
            return Ok();
        }

        public IActionResult ViaStartInfo() {
            string input = Request.Headers["X-Command"];
            // ruleid: csharp-os-command-injection
            var psi = new ProcessStartInfo("tool", input);
            // ruleid: csharp-os-command-injection
            Process.Start(psi);
            return Ok();
        }

        public IActionResult ViaProperty() {
            string file = Request.QueryString["file"];
            var p = new Process();
            // ruleid: csharp-os-command-injection
            p.StartInfo.FileName = file;
            p.Start();
            return Ok();
        }

        public IActionResult SafeConstant() {
            // ok: csharp-os-command-injection
            Process.Start("ping", "127.0.0.1");
            return Ok();
        }

        public IActionResult SafeLiteral() {
            var psi = new ProcessStartInfo("ping", "127.0.0.1");
            // ok: csharp-os-command-injection
            Process.Start(psi);
            return Ok();
        }
    }
}
